### PR TITLE
add movementSpeed attribute support

### PR DESCRIPTION
**Pull request by U5B**
attribute movementSpeed is used for sprinting and effect speed modification
sprinting and effect speed modification is server-side

Source Code from DecomplierMC - 1.16.4
```java
public float getSpeed() { // net.minecraft.world.entity.player.Player
  return (float)this.getAttributeValue(Attributes.MOVEMENT_SPEED);
}
private float getFrictionInfluencedSpeed(float f) { // net.minecraft.world.entity.LivingEntity
  if (this.onGround) {
      return this.getSpeed() * (0.21600002f / (f * f * f));
  }
  return this.flyingSpeed;
}
```

### CODE REVIEW

**Comment by U5B** [comment]
Marked as draft since at higher speeds (255) the bot goes faster than vanilla. (13 vs 10) in moved too quickly.

**Comment by U5B** [comment]
Alright I found an issue with negative acceleration causing the bot's velocity to go negative, so I set acceleration to 0 if it is negative.

To test the bot's speed is accurate enough, I did `console.log((vel.x + vel.z) * 20)` after moveEntity on L401 and compared it to Vanilla's speed per second on the mod MiniHud. Unfortunately, I could not find a more accurate speed counter then one to the hundredth.

**Comment by U5B** [comment]
Code referenced for client-side sprinting:
```java
public void setSprinting(boolean bl) {
    super.setSprinting(bl);
    AttributeInstance attributeInstance = this.getAttribute(Attributes.MOVEMENT_SPEED);
    if (attributeInstance.getModifier(SPEED_MODIFIER_SPRINTING_UUID) != null) {
        attributeInstance.removeModifier(SPEED_MODIFIER_SPRINTING);
    }
    if (bl) {
        attributeInstance.addTransientModifier(SPEED_MODIFIER_SPRINTING);
    }
}
```

**Comment by U5B** [comment]
https://cdn.discordapp.com/attachments/413438067471548427/919712091940732978/1.16.4_Forge_Velocity_Display.jar
Forge mod going to be used to compare prismarine-physics numbers to vanilla's numbers with different speed modifiers.

**Comment by U5B** [comment]
Waiting for https://github.com/PrismarineJS/minecraft-data/pull/486

**Comment by rom1504** [comment]
https://github.com/PrismarineJS/minecraft-data/pull/486 available

File: index.js [hunk 1]
```diff
@@ -378,12 +379,29 @@ function Physics (mcData, world) {
       let inertia = physics.airborneInertia
       const blockUnder = world.getBlock(pos.offset(0, -1, 0))
       if (entity.onGround && blockUnder) {
+        let playerAttributes
+        console.log(physics.movementSpeedAttribute)
```
Comment (u9g) [inline index.js:383]: should this be here?

File: lib/attribute.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+exports.getAttributeValue = function (prop) {
+  let X = prop.value
```
Comment (u9g) [inline lib/attribute.js:2]: best to not use capital letters for variable names

**Review by u9g — CHANGES_REQUESTED** [review]
Some nitpicks

File: index.js (hunk 1 above)
Comment (U5B) [reply index.js:383]: nope haha

File: lib/attribute.js (hunk 2 above)
Comment (U5B) [reply lib/attribute.js:2]: that was from Karang's code, I can fix

**Review by u9g — APPROVED** [review]
LGTM

File: index.js [hunk 3]
```diff
@@ -378,12 +379,28 @@ function Physics (mcData, world) {
...
       const blockUnder = world.getBlock(pos.offset(0, -1, 0))
       if (entity.onGround && blockUnder) {
+        let playerAttributes
+        if (entity.attributes && entity.attributes[physics.movementSpeedAttribute]) {
+          playerAttributes = entity.attributes[physics.movementSpeedAttribute]
+        } else {
+          playerAttributes = attribute.createAttributeValue(physics.playerSpeed) // default attribute
+        }
+        // Client-side sprinting (don't rely on server-side sprinting)
+        playerAttributes = attribute.deleteAttributeModifier(playerAttributes, '662a6b8d-da3e-4c1c-8813-96ea6097278d') // always delete sprinting (if it exists)
```
Comment (rom1504) [inline index.js:389]: what is that uuid

**Comment by rom1504** [comment]
code looks super magical
what can be done to make it clearer to the reader what's going on?

**Comment by U5B** [comment]
I looked at vanilla code and reproduced it here.
I hope this commit makes it more readable.

File: index.js (hunk 3 above)
Comment (U5B) [reply index.js]: SPEED_MODIFIER_SPRINTING_UUID is from LivingEntity.java

**Pull request merged**